Add support for --detach flag in stack deploy

gmargaritis opened the pull request:
Part of #373 along with https://github.com/docker/cli/pull/4259

Added `--detach` and `--quiet` / `-q` flags to `stack deploy`. Setting `--detach=false` waits until all of the services have converged. Shows progress bars for each individual task, unless  `--quiet` / `-q` is specified.

I have used the `WaitOnService` under the hood, so that the behavior is in line with `service update` or `service create` while using the `--detach` flag.

neersighted commented:
PTAL @corhere @dperny

gmargaritis commented:
Hey @thaJeztah and @neersighted 

I understand you're super busy, but wanted to see if there's anything I could do on my part to help move this forward. It will be a great improvement for our workflow, as we'll be able to deploy with confidence! Thanks!

marden commented:
Waiting for this too

thaJeztah commented on cli/command/stack/swarm/deploy_composefile.go:
```diff
@@ -256,10 +264,29 @@ func deployServices(ctx context.Context, dockerCli command.Cli, services map[str
-				return errors.Wrapf(err, "failed to create service %s", name)
+			response, err := apiClient.ServiceCreate(ctx, serviceSpec, createOpts)
+			if err != nil {
+				return errors.Wrapf(err, "Failed to create service %s", name)
```
These errors should probably remain lowercase (go convention it to have lowercase errors). (If needed, we can still add some prefix when printing)

gmargaritis replied:
Reverted it!

thaJeztah commented on cli/command/stack/swarm/deploy_composefile.go:
```diff
@@ -256,10 +264,29 @@ func deployServices(ctx context.Context, dockerCli command.Cli, services map[str
+	if !detach {
+		for _, serviceID := range serviceIDs {
+			if err := servicecli.WaitOnService(ctx, dockerCli, serviceID, quiet); err != nil {
+				errs = append(errs, fmt.Sprintf("%s: %v", serviceID, err))
```
haven't tried if it'd look good, and probably fine for a follow-up, but we may be able to use the new `errors.Join` here, and create a multi-error 🤔

gmargaritis replied:
@thaJeztah 👋 I have updated the code according to the suggestion by using `errors.Join()`.

Given the opportunity, I replaced `errors.Errorf` and `errors.Wrapf` with `fmt.Errorf` in this file, in order to avoid renaming the `errors` library.

gmargaritis commented:
Hey @thaJeztah and @neersighted 👋

Is there anything else I can do to make sure that this one moves forward?

thaJeztah commented on cli/command/stack/deploy.go:
```diff
@@ -42,5 +42,7 @@ func newDeployCommand(dockerCli command.Cli) *cobra.Command {
 	flags.StringVar(&opts.ResolveImage, "resolve-image", swarm.ResolveImageAlways,
 		`Query the registry to resolve image digest and supported platforms ("`+swarm.ResolveImageAlways+`", "`+swarm.ResolveImageChanged+`", "`+swarm.ResolveImageNever+`")`)
 	flags.SetAnnotation("resolve-image", "version", []string{"1.30"})
+	flags.BoolVarP(&opts.Detach, "detach", "d", true, "Exit immediately instead of waiting for the stack services to converge")
+	flags.BoolVarP(&opts.Quiet, "quiet", "q", false, "Suppress progress output")
```
It looks like this PR (currently at least) keeps the old behavior (`docker stack deploy` being non-interactive / detached as default).

If the plan is to eventually make "interactive" (`--detach=false`) the default, I also went looking how we handled the transition for `docker service create`; https://github.com/moby/moby/commit/21ec12b9676f95aaa69c3082b97d0b10031cc90a (https://github.com/moby/moby/pull/31144)

For that we chose to print a warning; wondering if (and if: when? / which release?) we want to do the same for `docker stack`; inform users that we plan to switch non-detach as default, with the option to return to the old (`--detach[=true]`) behavior.

gmargaritis replied:
I've added a message just like in [moby/moby@21ec12b](https://github.com/moby/moby/commit/21ec12b9676f95aaa69c3082b97d0b10031cc90a) ([moby/moby#31144](https://github.com/moby/moby/pull/31144)) as you said, indicating that we plan to switch non-detach as default.

gmargaritis commented on cli/command/stack/swarm/deploy_composefile.go:
```diff
@@ -169,13 +171,13 @@ func createNetworks(ctx context.Context, dockerCli command.Cli, namespace conver
 }
 
-func deployServices(ctx context.Context, dockerCli command.Cli, services map[string]swarm.ServiceSpec, namespace convert.Namespace, sendAuth bool, resolveImage string) error {
+func deployServices(ctx context.Context, dockerCli command.Cli, services map[string]swarm.ServiceSpec, namespace convert.Namespace, sendAuth bool, resolveImage string, detach bool, quiet bool) error {
```
Hey @thaJeztah 👋,

Passing `options.Deploy` in `deployServices` was increasing the cyclomatic complexity. I've decided to keep `deployServices` as it previously was and call `waitOnServices` one level higher. This way the logic is clearer while keeping the cyclomatic complexity as it was.

https://github.com/gmargaritis/cli/blob/e22a17dc65a56dd1f6b5ba01e0f440049c04482b/cli/command/stack/swarm/deploy_composefile.go#L65-L74

gmargaritis commented:
Hey @thaJeztah @neersighted 👋

I've implemented all the suggested fixes and rebased the PR!

Is there anything else that I can do to move this one forward?

gmargaritis commented:
@thaJeztah It seems that we have to re-run the ci since it failed because of an error not related with the PR. 

Let me know if there's anything left to be done here ✌️

thaJeztah commented:
I gave it a kick.

For posterity; this was the error. Probably flaky, or some connection issues on GitHub actions (not sure if I've seen that specific one before, so cc @crazy-max - just in case it pops-up more frequently)

```
Dockerfile:1
--------------------
   1 | >>> # syntax=docker/dockerfile:1
   2 |     
   3 |     ARG BASE_VARIANT=alpine
--------------------
ERROR: failed to solve: frontend grpc server closed unexpectedly
Error: buildx bake failed with: ERROR: failed to solve: frontend grpc server closed unexpectedly
```

gmargaritis commented:
@thaJeztah @crazy-max 

What's the status of this PR? Are we ok here?

laurazard commented on cli/command/stack/swarm/deploy_composefile.go:
```diff
@@ -257,10 +272,29 @@ func deployServices(ctx context.Context, dockerCli command.Cli, services map[str
 		}
 	}
+
+	if len(errs) > 0 {
+		return errors.Join(errs...)
+	}
```
I _think_ you can just return `errors.Join(errs...)` here, it'll be `nil` if `len(errs) == 0`

laurazard reviewed (approved):
Overall LGTM, thank you!

thaJeztah commented:
Did a quick rebase and re-generated the markdown as we updated the `cli-docs-tool` with a fix for the boolean flags;

```patch
diff --git a/docs/reference/commandline/stack_deploy.md b/docs/reference/commandline/stack_deploy.md
index c8b116a09..82934f6f9 100644
--- a/docs/reference/commandline/stack_deploy.md
+++ b/docs/reference/commandline/stack_deploy.md
@@ -12,7 +12,7 @@ Deploy a new stack or update an existing stack
 | Name                                                     | Type          | Default  | Description                                                                                       |
 |:---------------------------------------------------------|:--------------|:---------|:--------------------------------------------------------------------------------------------------|
 | [`-c`](#compose-file), [`--compose-file`](#compose-file) | `stringSlice` |          | Path to a Compose file, or `-` to read from stdin                                                 |
-| `-d`, `--detach`                                         |               |          | Exit immediately instead of waiting for the stack services to converge                            |
+| `-d`, `--detach`                                         | `bool`        | `true`   | Exit immediately instead of waiting for the stack services to converge                            |
 | `--prune`                                                |               |          | Prune services that are no longer referenced                                                      |
 | `-q`, `--quiet`                                          |               |          | Suppress progress output                                                                          |
 | `--resolve-image`                                        | `string`      | `always` | Query the registry to resolve image digest and supported platforms (`always`, `changed`, `never`) |
```


I also opened a tracking ticket to look at follow-up improvements discussed above; https://github.com/docker/cli/issues/4907

thaJeztah reviewed (approved):
LGTM let's bring this one in as a first iteration.
